core/src/main/java/ai/timefold/solver/core/impl/bavet/common/AbstractPrecomputeNode.java

```diff
@@ -0,0 +1,184 @@
+        implements BavetRootNode<Object> {
+    private final StaticPropagationQueue<Tuple_> propagationQueue;
+    private final Map<Object, List<Tuple_>> objectToOutputTuplesMap = new IdentityHashMap<>(1000);
+    private final IdentityHashMap<Tuple_, Tuple_> inputTupleToOutputTupleMap = new IdentityHashMap<>(1000);
+    private final NodeNetwork innerNodeNetwork;
+    private final RecordingTupleLifecycle<Tuple_> recordingTupleNode;
+    private final Class<?>[] sourceClasses;
+    private final Set<Object> queuedInsertSet = CollectionUtils.newIdentityHashSet(32);
+    private final Set<Object> queuedUpdateSet = CollectionUtils.newIdentityHashSet(32);
+    private final Set<Object> queuedRetractSet = CollectionUtils.newIdentityHashSet(32);
+
+    protected AbstractPrecomputeNode(NodeNetwork innerNodeNetwork,
+            RecordingTupleLifecycle<Tuple_> recordingTupleNode,
+            TupleLifecycle<Tuple_> nextNodesTupleLifecycle,
+            Class<?>[] sourceClasses) {
+        this.innerNodeNetwork = innerNodeNetwork;
+        this.propagationQueue = new StaticPropagationQueue<>(nextNodesTupleLifecycle);
+        this.sourceClasses = sourceClasses;
+        this.recordingTupleNode = recordingTupleNode;
+    }
+
+    @Override
+    public final Propagator getPropagator() {
+        return propagationQueue;
+    }
+
+    @Override
+    public final boolean allowsInstancesOf(Class<?> clazz) {
+        for (var sourceClass : sourceClasses) {
+            if (sourceClass.isAssignableFrom(clazz)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public final Class<?>[] getSourceClasses() {
+        return sourceClasses;
+    }
+
+    @Override
+    public final boolean supports(BavetRootNode.LifecycleOperation lifecycleOperation) {
+        return true;
+    }
+
+    @Override
+    public final void insert(Object a) {
+        // do not remove a retract of the same fact (a fact was updated)
+        queuedInsertSet.add(a);
+    }
+
+    @Override
+    public final void update(Object a) {
+        queuedUpdateSet.add(a);
+    }
+
+    private void updateExisting(@Nullable Object a, Tuple_ tuple) {
+        var state = tuple.state;
+        if (state.isDirty()) {
+            if (state == TupleState.DYING || state == TupleState.ABORTING) {
+                throw new IllegalStateException("The fact (%s) was retracted, so it cannot update."
+                        .formatted(a));
+            }
+            // CREATING or UPDATING is ignored; it's already in the queue.
+        } else {
+            propagationQueue.update(tuple);
+        }
+    }
+
+    @Override
+    public final void retract(Object a) {
+        // remove an insert then retract (a fact was inserted but retracted before settling)
+        // do not remove a retract then insert (a fact was updated)
+        if (!queuedInsertSet.remove(a)) {
```

Comment:
Possibly incomplete? If we see a retract, we arguably want to remove any update as well.